src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.SymbolToProjectId.cs

```diff
@@ -51,9 +53,24 @@ internal partial class SolutionState
             {
                 if (!unrootedSymbolToProjectId.TryGetValue(symbol, out var projectId))
                 {
+                    // look through all the projects, and see if this symbol came from the primary assembly for that
+                    // project.  (i.e. was a source symbol from that project, or was retargetted into that project).  If
+                    // so, that's the originating project.
                     foreach (var (id, tracker) in _projectIdToTrackerMap)
                     {
-                        if (tracker.ContainsAssemblyOrModuleOrDynamic(symbol))
+                        if (tracker.ContainsAssemblyOrModuleOrDynamic(symbol, primary: true))
+                        {
+                            projectId = id;
+                            break;
+                        }
+                    }
+
+                    // now, look through all the secondary assemblies (i.e. references) for that project.  This is the
+                    // case for metadata symbols.  A metadata symbol might be found in many projects, so we just return
+                    // the first result as that's just as good for finding the metadata symbol as any other project.
+                    foreach (var (id, tracker) in _projectIdToTrackerMap)
+                    {
+                        if (tracker.ContainsAssemblyOrModuleOrDynamic(symbol, primary: false))
```

Comment:
here we walk thr trackers twice.  we always try to map any symbol to the true originating project for it.  And only if we don't have that, do we then do the secondary pass, where we'll accept the project even if it contains the symbol as a reference.